funcx_endpoint/funcx_endpoint/endpoint/rabbit_mq/result_queue_subscriber.py

```diff
@@ -0,0 +1,268 @@
+            self.queue_name, on_message_callback=self.on_message, exclusive=True
+        )
+
+    def on_consumer_cancelled(self, method_frame):
+        """Invoked by pika when RabbitMQ sends a Basic.Cancel for a consumer
+        receiving messages.
+
+        :param pika.frame.Method method_frame: The Basic.Cancel frame
+
+        """
+        logger.info("Consumer was cancelled remotely, shutting down: %r", method_frame)
+        if self._channel:
+            self._channel.close()
+
+    def on_message(self, _unused_channel, basic_deliver, _properties, body):
+        """on_message pushes a message upon receipt into the external_queue
+        for async consumption. The message pushed to the external_queue
+        is of the following type : Tuple[str, bytes]
+
+        :param pika.channel.Channel unused_channel: The channel object
+        :param pika.Spec.Basic.Deliver: basic_deliver method
+        :param pika.Spec.BasicProperties: properties
+        :param str|unicode body: The message body
+
+        """
+
+        routing_key = basic_deliver.routing_key
+
+        if self.external_queue:
+            try:
+                response: Tuple[str, bytes] = (routing_key.rsplit(".", 1)[0], body)
+                self.external_queue.put(response)
+            except queue.Full:
+                logger.exception("Failed to forward message to external_queue")
+                self._channel.basic_nack(basic_deliver.delivery_tag, requeue=True)
+            else:
+                self._channel.basic_ack(delivery_tag=basic_deliver.delivery_tag)
+
+    def stop_consuming(self):
+        """Tell RabbitMQ that you would like to stop consuming by sending the
+        Basic.Cancel RPC command.
+
+        """
+        if self._channel:
+            logger.info("Sending a Basic.Cancel RPC command to RabbitMQ")
+            self._channel.basic_cancel(self.on_cancelok, self._consumer_tag)
+
+    def on_cancelok(self, unused_frame):
+        """This method is invoked by pika when RabbitMQ acknowledges the
+        cancellation of a consumer. At this point we will close the channel.
+        This will invoke the on_channel_closed method once the channel has been
+        closed, which will in-turn close the connection.
+
+        :param pika.frame.Method unused_frame: The Basic.CancelOk frame
+
+        """
+        logger.info("RabbitMQ acknowledged the cancellation of the consumer")
+        self._channel.close()
+
+    def run(self):
+        """Run the example consumer by connecting to RabbitMQ and then
+        starting the IOLoop to block and allow the SelectConnection to operate.
+
+        """
+        self._connection = self._connect()
+        self._connection.ioloop.start()
+
+    def close(self) -> None:
+        """Close the connection"""
+        logger.info("Attempting connection close")
+        self._closing = True
+        self.stop_consuming()
+        self._connection.close()
+        logger.info("Connection closed")
```

Comment:
Same as the other subscriber class, if this is going to use the `close()` name, it needs to call `super().close()`.

Reply:
If we need terminate to work, then we need to do all of the work of setting up the sigterm handler.

Is `close()` actually failing in the tests, or is it just slow because it waits for the `join()`?
I don't think we should be killing things in our tests in a mode that is different from the one we want in production.